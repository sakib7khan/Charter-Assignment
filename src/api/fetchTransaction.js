export const fetchTransactions = async () => {
  const res = await fetch('http://localhost:3005/transactions');
  if (!res.ok) throw new Error('API fetch failed');
  return await res.json();
};
