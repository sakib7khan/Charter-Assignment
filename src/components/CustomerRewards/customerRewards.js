// customerRewards.js

import React, { useEffect, useState } from 'react';
import PropTypes from 'prop-types';
import { fetchTransactions } from '../../api/fetchTransaction';
import { calculateRewards } from '../../utils/calculateRewards';
import { Container, Title, Table, Spinner, ErrorMessage } from '../../styles/stylesComponents';

const CustomerRewards = ({ customerId, selectedMonth, selectedYear }) => {
  const [monthlyRewards, setMonthlyRewards] = useState({});
  const [totalRewards, setTotalRewards] = useState(0);
  const [loading, setLoading] = useState(true);
  const [error, setError] = useState(null);

  useEffect(() => {
    fetchTransactions()
      .then((data) => {
        const filtered = data.filter((txn) => txn.customerId === customerId);
        const rewardsPerMonth = {};
        let total = 0;

        filtered.forEach((txn) => {
          const date = new Date(txn.date);
          const month = date.toLocaleString('default', { month: 'short' });
          const year = date.getFullYear();

          if (year !== parseInt(selectedYear)) return;

          const key = `${month}-${year}`;
          const rewards = calculateRewards(txn.amount);

          if (!rewardsPerMonth[key]) rewardsPerMonth[key] = 0;
          rewardsPerMonth[key] += rewards;
          total += rewards;
        });

        setMonthlyRewards(rewardsPerMonth);
        setTotalRewards(total);
        setLoading(false);
      })
      .catch((err) => {
        setError('Error loading rewards.');
        setLoading(false);
      });
  }, [customerId, selectedYear]);

  if (loading) return <Spinner data-testid="loading-spinner">Loading rewards...</Spinner>;
  if (error) return <ErrorMessage data-testid="error-message">{error}</ErrorMessage>;

  return (
    <Container data-testid="customer-rewards-container">
      <Title data-testid="rewards-title">Rewards Summary for {customerId}</Title>
      <h3 data-testid="total-rewards">Total Rewards: {totalRewards}</h3>
      <Table data-testid="rewards-table">
        <thead>
          <tr>
            <th>Month</th>
            <th>Rewards</th>
          </tr>
        </thead>
        <tbody>
          {Object.entries(monthlyRewards).map(([monthKey, points]) => (
            <tr key={monthKey} data-testid={`reward-row-${monthKey}`}>
              <td data-testid={`month-${monthKey}`}>{monthKey}</td>
              <td data-testid={`points-${monthKey}`}>{points}</td>
            </tr>
          ))}
        </tbody>
      </Table>
    </Container>
  );
};

CustomerRewards.propTypes = {
  customerId: PropTypes.string.isRequired,
  selectedMonth: PropTypes.string.isRequired,
  selectedYear: PropTypes.string.isRequired
};

export default CustomerRewards;
