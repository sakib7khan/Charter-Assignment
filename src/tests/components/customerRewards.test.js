import React from 'react';
import { render, screen, waitFor } from '@testing-library/react';
import CustomerRewards from '../../components/CustomerRewards/customerRewards';
import * as fetchAPI from '../../api/fetchTransaction';
import '@testing-library/jest-dom';

// Mock data
const mockTransactions = [
  { transactionId: 'T1', customerId: 'C001', amount: 120, date: '2025-01-15' },
  { transactionId: 'T2', customerId: 'C001', amount: 80, date: '2025-01-20' },
  { transactionId: 'T3', customerId: 'C001', amount: 50, date: '2025-02-10' },
];

// Utility mock
jest.mock('../../api/fetchTransaction');

describe('CustomerRewards Component', () => {
  afterEach(() => jest.clearAllMocks());

  const customerId = 'C001';
  const selectedYear = '2025';
  const selectedMonth = 'Jan';

  it('renders loading spinner initially', async () => {
    fetchAPI.fetchTransactions.mockReturnValue(new Promise(() => {}));

    render(
      <CustomerRewards
        customerId={customerId}
        selectedMonth={selectedMonth}
        selectedYear={selectedYear}
      />
    );
    expect(screen.getByTestId('loading-spinner')).toBeInTheDocument();
  });

  it('renders error message on API failure', async () => {
    fetchAPI.fetchTransactions.mockRejectedValueOnce(new Error('Failed'));

    render(
      <CustomerRewards
        customerId={customerId}
        selectedMonth={selectedMonth}
        selectedYear={selectedYear}
      />
    );
    await waitFor(() =>
      expect(screen.getByTestId('error-message')).toBeInTheDocument()
    );
    expect(screen.getByTestId('error-message')).toHaveTextContent(
      'Error loading rewards.'
    );
  });

  it('renders correct total and table rows', async () => {
    fetchAPI.fetchTransactions.mockResolvedValueOnce(mockTransactions);

    render(
      <CustomerRewards
        customerId={customerId}
        selectedMonth={selectedMonth}
        selectedYear={selectedYear}
      />
    );

    await waitFor(() =>
      expect(
        screen.getByTestId('customer-rewards-container')
      ).toBeInTheDocument()
    );

    // Check Title
    expect(screen.getByTestId('rewards-title')).toHaveTextContent(
      `Rewards Summary for ${customerId}`
    );

    // Check Total Rewards
    const reward1 = 90; // 120 -> 50 over 50 = 50 pts + 2*20 = 40 => 90
    const reward2 = 30; // 80 -> 50 over 50 = 30
    const reward3 = 0; // 50 -> No reward

    expect(screen.getByTestId('total-rewards')).toHaveTextContent(
      `Total Rewards: ${reward1 + reward2}`
    );

    // Table rows
    expect(screen.getByTestId('month-Jan-2025')).toBeInTheDocument();
    expect(screen.getByTestId('points-Jan-2025')).toHaveTextContent(
      reward1 + reward2
    );

    expect(screen.getByTestId('month-Feb-2025')).toBeInTheDocument();
    expect(screen.getByTestId('points-Feb-2025')).toHaveTextContent('0');
  });
  it('matches the snapshot', () => {
    fetchAPI.fetchTransactions.mockResolvedValueOnce(mockTransactions);
    const { asFragment } = render(<CustomerRewards />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('does not include transactions from other years', async () => {
    fetchAPI.fetchTransactions.mockResolvedValueOnce([
      ...mockTransactions,
      {
        transactionId: 'T4',
        customerId: 'C001',
        amount: 200,
        date: '2024-12-01',
      }, // Should be ignored
    ]);

    render(
      <CustomerRewards
        customerId={customerId}
        selectedMonth={selectedMonth}
        selectedYear='2025'
      />
    );
    await waitFor(() =>
      expect(
        screen.getByTestId('customer-rewards-container')
      ).toBeInTheDocument()
    );

    expect(screen.queryByText('Dec-2024')).not.toBeInTheDocument();
  });
});
