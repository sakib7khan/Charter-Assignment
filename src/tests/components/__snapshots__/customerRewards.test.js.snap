// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CustomerRewards Component matches the snapshot 1`] = `
<DocumentFragment>
  <div
    class="sc-iQQCXo bezgzY"
    data-testid="loading-spinner"
  >
    Loading rewards...
  </div>
</DocumentFragment>
`;
